Update to error messages, and help text

jordenrabasco opened the pull request:
- this PR is to address issue #87 
- adds additional text in the descriptions of the params that indicate which methods they should be used with
- error catching framework expanded to include which params should be used with which decontam methods

jordenrabasco committed:
Message updates

- added text in plugin.py to better inform user of which params go with which method.
-Added additional valueerrors to decontam.py to check for missing params

jordenrabasco committed:
clean up

jordenrabasco committed:
fixed error

- fixed mistype error

colinvwood commented on q2_quality_control/decontam.py:
```diff
@@ -14,36 +14,49 @@
+                         'please check that input parameters'
+                         ' --prev-control-column and '
+                         '--p-prev-control-indicator are being utilized')
     else:
```
These `else` statements aren't necessary because of the `ValueError`s being raised. Removing these will also help with nesting.

jordenrabasco replied:
addressed in commit [f31210b](https://github.com/qiime2/q2-quality-control/pull/88/commits/f31210b1b4524e53c8545435910ecb106d40fc8d)

colinvwood commented on q2_quality_control/decontam.py:
```diff
@@ -14,36 +14,49 @@
+                                               .unique())))
             else:
-                print("All appropriate inputs are found")
+                prev_controls = metadata.loc[
+                    metadata[prev_control_column] == prev_control_indicator]
+                prev_sample_names = prev_controls.index.values
+                indic = 0
+                for name in prev_sample_names:
+                    if name in table.index.values:
+                        indic = indic + 1
+                if indic < 5:
+                    print("We recommend 5 control samples - " +
+                          str(indic) + " found")
```
For clarity, let's try to refactor this to use some common pandas methods, specifically `Index.intersection` and `Series.sum` will be useful.

jordenrabasco replied:
addressed in commit [f31210b](https://github.com/qiime2/q2-quality-control/pull/88/commits/f31210b1b4524e53c8545435910ecb106d40fc8d)

colinvwood commented on q2_quality_control/decontam.py:
```diff
@@ -14,36 +14,49 @@
+                         'that input parameter'
+                         ' --p-freq-concentration-column is '
+                         'being utilized')
+    else:
```
Delete for same reason as above

jordenrabasco replied:
addressed in commit [f31210b](https://github.com/qiime2/q2-quality-control/pull/88/commits/f31210b1b4524e53c8545435910ecb106d40fc8d)

colinvwood commented on q2_quality_control/decontam.py:
```diff
@@ -58,27 +71,23 @@ def _check_column_inputs(table, metadata, method, freq_concentration_column,
-                                    prev_control_indicator)
-
+        if freq_concentration_column is not None:
+            print("Warning: --p-freq-concentration-column given, but" +
```
You'll want to use the `warnings` module for warnings instead of print so that the warnings status as a warning is clear (and presented to a user in all cases). `print` sends to a different output stream and is ignored in qiime2 without the `--verbose` flag. 

If you don't want to worry about all that it could make sense to simply error on unexpected inputs so that it's made very explicit to users that the parameter is not being taken into account.

jordenrabasco replied:
I opted for just raising a value error instead of the warning as to avoid confusion with users in the future about what parameters are being utilized. Addressed in commit [f31210b](https://github.com/qiime2/q2-quality-control/pull/88/commits/f31210b1b4524e53c8545435910ecb106d40fc8d)

colinvwood commented on q2_quality_control/decontam.py:
```diff
@@ -58,27 +71,23 @@ def _check_column_inputs(table, metadata, method, freq_concentration_column,
-        else:
-            print("All appropriate inputs are found")
+        if prev_control_column is not None:
+            print("Warning: --prev-control-column given, but" +
```
Same as above

jordenrabasco replied:
addressed in commit [f31210b](https://github.com/qiime2/q2-quality-control/pull/88/commits/f31210b1b4524e53c8545435910ecb106d40fc8d)

colinvwood commented on q2_quality_control/decontam.py:
```diff
@@ -58,27 +71,23 @@ def _check_column_inputs(table, metadata, method, freq_concentration_column,
+            print("Warning: --prev-control-column given, but" +
+                  " disregarded")
+        if prev_control_indicator is not None:
+            print("Warning: --p-prev-control-indicator given, but" +
```
Same as above

jordenrabasco replied:
addressed in commit [f31210b](https://github.com/qiime2/q2-quality-control/pull/88/commits/f31210b1b4524e53c8545435910ecb106d40fc8d)

colinvwood commented on q2_quality_control/decontam.py:
```diff
@@ -104,6 +113,7 @@ def decontam_identify(table: pd.DataFrame,
     metadata = metadata.to_dataframe()
     _check_column_inputs(table, metadata, method, freq_concentration_column,
                          prev_control_column, prev_control_indicator)
+    print("All approriate inputs found")
```
We don't need this; the action completing implies this

jordenrabasco replied:
addressed in commit [f31210b](https://github.com/qiime2/q2-quality-control/pull/88/commits/f31210b1b4524e53c8545435910ecb106d40fc8d)

colinvwood reviewed (commented):
It would be best if we could add some tests for these changes, do you think that's possible @jordenrabasco?

jordenrabasco committed:
requested changes and tests

-adds the requested changes to the PR and the tests for the various valueerrors

colinvwood commented on q2_quality_control/tests/test_decontam.py:
```diff
@@ -115,6 +115,63 @@ def test_combined(self):
             self.assertEqual(test_table, expecter_table)
 
+    def test_freq_prev_control_col_error_raised(self):
+        with self.assertRaises(ValueError) as context:
```
There's a method called `assertRaisesRegex` that takes two arguments, the error that is expected to be raised, and a regex string that is expected to match the error message. For example, in this case you could use:

```python
with self.assertRaisesRegex(ValueError, '--prev-control-column'):
    ...
```

and get the same functionality as the `assertRaises` and `assertIn` assertions combined. (I think your approach is very clear as well, so no need to refactor, but just thought I'd let you know.)

jordenrabasco replied:
Thanks for the info!

colinvwood commented on q2_quality_control/tests/test_decontam.py:
```diff
@@ -115,6 +115,63 @@ def test_combined(self):
+                method='frequency',
+                freq_concentration_column='quant_reading',
+                prev_control_column='Sample_or_Control')
+        self.assertIn("--prev-control-column", str(context.exception))
```
Because the string `"--prev-control-column"` occurs in more than one error message it doesn't uniquely identify any error. You should choose the asserted substring such that it uniquely identifies the error being tested.

colinvwood replied:
I won't leave separate comments for the other tests, but make to sure to update them in this way as well.

jordenrabasco replied:
addressed in commit addressed in commit 82acad4b955d831983ba4ee95b4450d0f1109e4b

colinvwood commented on q2_quality_control/tests/test_decontam.py:
```diff
@@ -115,6 +115,63 @@ def test_combined(self):
+                method='frequency')
+        self.assertIn("--p-freq-concentration-column", str(context.exception))
+
+    def test_prev_params_no_indic_error_raised(self):
```
This method name says that we are testing that a certain type of error does not get raised but the method itself asserts that an error is raised.

jordenrabasco replied:
ah okay I changed the name of the test to be more clear, addressed in commit 82acad4b955d831983ba4ee95b4450d0f1109e4b

colinvwood commented on q2_quality_control/tests/test_decontam.py:
```diff
@@ -115,6 +115,63 @@ def test_combined(self):
+                prev_control_column='Sample_or_Control')
+        self.assertIn("--prev-control-column ", str(context.exception))
+
+    def test_prev_params_no_col_error_raised(self):
```
Same as line 146

jordenrabasco replied:
same as above

jordenrabasco committed:
Merge branch 'qiime2:dev' into dev

jordenrabasco committed:
testing update

-updates testing names to be more clear
-updates error message check to be more unique

colinvwood commented:
Thank you @jordenrabasco!
